### ISSUE TITLE
[Linux] Cursor not locked in Window/Mouse not working

### ISSUE THREAD

**Issue by smaTc** [Try/OpenGothic#82]
On a dualscreen setup the cursor is not locked correctly on the primary display/the display running on in the window and therefore the mouse input/camera is not working

**Comment by Try** [comment]
Hi, @smaTc and thanks for bug report!

Marked as help-need, because I don't have a dualscreen.

**Comment by smaTc** [comment]
Well my C++ Skills are very limited right now, but if you tell me where you handle the cursor and window creation etc I will try to debug it.

**Comment by Try** [comment]
By design window management is implemented on graphics-engine side
for Linux: OpenGothic/lib/MoltenTempest/Engine/system/api/x11api.cpp

```c++
void X11Api::implSetCursorPosition(int x, int y) {
  activeCursorChange = 1;
  XWarpPointer(dpy, None, root, 0, 0, 0, 0, x, y); // <-- issue can be here, because cursor is binded to root aka DefaultRootWindow
  XFlush(dpy);
  }
```

**Comment by smaTc** [comment]
It seems the problem is that you warp the cursor to the screen rootWindow and not to the focused window. I am not completely sure if that really fixed it, but the cursor disappeared/was locked inside the game window and I could control the camera via mouse. There could also be more cases where the wrong window is addressed. But could also be random that this solution works.

```
void X11Api::implSetCursorPosition(int x, int y) {
  activeCursorChange = 1;
 
  //that fixed it
  ::Window w;
  int revert_to;
  XGetInputFocus(dpy, &w, &revert_to); 

  XWarpPointer(dpy, None, w , 0, 0, 0, 0, x, y);
  XFlush(dpy);
  }
```

**Comment by Try** [comment]
Thanks @smaTc for looking into issue.
One think I don't like about solution based on `XGetInputFocus` is fact that game window may not be focused.

My idea is to add additional window parameter to setCursorPosition, that way XGetInputFocus will not be necessary

**Comment by Try** [comment]
In my path, I left X11 implementation unchanged for now, can you  please check, if code bellow gonna work:
```c++
void X11Api::implSetCursorPosition(SystemApi::Window *w, int x, int y) {
  activeCursorChange = 1;
  XWarpPointer(dpy, None, HWND(w), 0, 0, 0, 0, x, y);
  XFlush(dpy);
  }
```

**Comment by smaTc** [comment]
That code lets the cursor disappear, so it seems to fix that issue. But there seems to be some issue with window focus and mouse input in general somewhere (at least on linux, haven't tested windows), for example tabbing out does not work well. And is there a reason why mouse camera control is so crazy fast?

**Comment by Try** [comment]
> But there seems to be some issue with window focus

Must be something linux-related, at least on windows alt+tab works without problem

> why mouse camera control is so crazy fast?

Fast is relative term, so there is 2 cases:
1. fast = fast for you
2. fast = fast, because of bug

Speed for mouse-camera is fixed at this moment at 0.4 degree per pixel:
```c++
//playercontrol.cpp
void PlayerControl::onRotateMouse(int dAngle) {
  dAngle = std::max(-100,std::min(dAngle,100));
  rotMouse += float(dAngle)*0.4f;
  }
```
